README.md
# UptoDate

A tiny time and date helper 'library'

[![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](https://opensource.org/licenses/MIT)

### Features
* Clone Date
    * immutable
* Add / Subtract
    * Years, Months, Weeks, Days, Hours, Minutes, Seconds, Milliseconds
* Local Date
    * eg: Sunday, 30 June 2019
* Local Time
    * eg: 20:38:48
* Get Day 
    * eg: Sunday
* Get Month
    * eg: June
* Get Day of Year
    * eg 179
* Format Date
    * eg: 30/6/2019
    * eg: 30/6/19
    * eg: Sunday 30 June 2019
* Format Time
* Set Locale
    eg: 'en-US'
* Plugin System
    * extend uptoDate with your own custom plugins
    * see the plugin folder for examples
* ≈3kb minified
    * without plugins
* ≈1kb minifed & gzipped

### Demo

[https://retrovx.github.io/uptoDate](https://retrovx.github.io/uptoDate)

## Getting Started

### Install

```
git clone https://github.com/RetroVX/uptoDate.git
```
Or download from Zip

Then serve on your favourite local web server :)  

HTML  
```html
<script type="text/javascript" src="path/to/uptoDate.min.js"></script>
```

Javascript
```javascript
const utDate = new uptoDate();
```

## Examples

### Clone Date
```javascript
utDate.cloneDate(date);

// Example:
// Dates are mutable so if we copy the date variable then edit the copy variable, it changes the date variable!
// mutable
const date = new Date(2019, 5, 30, 20, 38);
const copy = date;
// add 5 hours to copy
copy.setHours(copy.getHours() + 5);
// output
console.log(date);
'Mon Jul 01 2019 01:38:00 GMT+0100 (British Summer Time)'
console.log(copy);
'Mon Jul 01 2019 01:38:00 GMT+0100 (British Summer Time)'

// immutable way with cloneDate()
const date = new Date(2019, 5, 30, 20, 38);
const copy = utDate.cloneDate(date);
// add 5 hours to copy
copy.setHours(copy.getHours() + 5);
// output
console.log(date);
'Sun Jun 30 2019 20:38:00 GMT+0100 (British Summer Time)'
console.log(copy);
'Mon Jul 01 2019 01:38:00 GMT+0100 (British Summer Time)'
```

### Modify Years, Months, Weeks, Days, Hours, Minutes, Seconds Or Milliseconds

#### Add
```javascript
// date is optional
utDate.add('hours', 3, date);
// Sun Jun 30 2019 23:38:00 GMT+0100 (British Summer Time)
```

#### Subtract
```javascript
// date is optional
utDate.subtract('years', 3, date);
// Sun Jun 30 2016 20:38:00 GMT+0100 (British Summer Time)
```

#### String Inputs For Add/Subtract
```javascript
'years', 'year', 'months', 'month', 'weeks', 'week' 'day', 'days', 'hours', 'hour', 'minutes', 'minute', 'seconds' 'second', 'milliseconds', 'millisecond'
```

### Get Date
```javascript
// add optional date, locale ('en-US') and/or timezone 
utDate.getLocalDate(date, locale, timezone);
// Sunday, 30 June 2019
```

### Get Time
```javascript
// add optional date, locale ('en-US') and/or timezone 
utDate.getLocalTime(date, locale, timezone);
// 20:38:48
```

### Get Day
```javascript
// add optional date & short length
utDate.formatDay(date, false);
// Sunday || Sun
```

### Get Month
```javascript
// add optional date & short length
utDate.formatMonth(date, false);
// June || Jun
```

### Get Day of Year
```javascript
// enter date
utDate.getDayOfYear(date);
// 179
```

### Get Days in a Month
```javascript
utDate.getDaysInMonth(date);
// 30
```

### Format Date
```javascript
// add optional date
utDate.formatDate(date);
// returns object
{   
    full: 'Sunday, 30 June 2019',
    day: 30, 
    fullDay: 'Sunday', 
    dayOfYear: 179,
    daysInMonth: 30,
    shortDay: 'Sun',
    month: 6, 
    shortMonth: 'Jun', 
    fullMonth: 'June', 
    year: 2019, 
    shortYear: 19,
    iso: '2019-06-30T20:38:00.000Z',
    utc: 'Sun, 30 Jun 2019 20:38:00 GMT'
}
```

### Format Time
```javascript
// add optional date
utDate.formatTime(date);
// returns object
{
    hours: 20,
    minutes: 38,
    seconds: 48,
    ms: 1562011912273
}
```

### Set Locale
```javascript
utDate.setLocale(locale);

// setLocale is chainable
utDate.setLocale('en-US').getLocalDate();
```

### Get Time Difference
```javascript
// input start time and end time using Date() format;
utDate.getTimeDifference(time1, time2);

// Example

const start = new Date();
// after 5 seconds get the time difference between start and date variables
setTimeout(function(){
    const date = new Date();
    utDate.timeDifference(start, date);
    // returns object with days, hours, minutes and seconds parameters
    // this would output { days: 0, hours: 0, minutes: 0, seconds: 5}
}, 5000);
```

### Add Plugin
Look inside the plugins folder to see current plugins. Each one will have a README attached.

HTML
```html
<!-- Make sure the plugin is added after uptoDate.js-->
<script type="text/javascript" src="path/to/plugin.js"></script>
```
Javascript
```javascript
// options is optional
utDate.addPlugin(yourPluginName, options);

// multiple plugins
utDate.addPlugin([yourPluginName, anotherPlugin], options);
```

#### Plugin Template
Look inside the plugin folder to find a plugin template :)

### Basic Example
```javascript
const utDate = new uptoDate();
utDate.addPlugin(countdownPlugin);

const start = new Date();
const countdownTo = utDate.add('minutes', 5);

// run every second
setInterval(function(){
    // the countdown will complete in 5 minutes
    utDate.countdown(countdownTo);

    // time Ago, returns time since started
    utDate.timeAgo(start);
}, 1000);
```

### Pre 1.0.0
Expect breaking changes to the api while under version 1.0.0

### Version 0.8

CHANGELOG.md
UptoDate Changelog

### Version 0.8

* new function ```getDaysInMonth``` that returns the number days in a month
* ```formatDate```
    * replaced getFullYear property with year
    * added daysInMonth property

### Version 0.7.1

* new plugin ```startOfPlugin```
    * ```startOf```
    * ```startOfDay```
    * ```startOfWeek```
    * ```startOfMonth```
    * ```startOfYear```

### Version 0.7.0

* moved ```isToday``` into the ```queryPlugin```
* moved ```timeAgo``` and ```timeIn``` into a new plugin called ```relativeTimePlugin```
* new plugin ```queryPlugin```
    * ```isToday```
    * ```isYesterday```
    * ```isTomorrow```
    * ```isWeekend```
    * ```isPast```
    * ```isFuture```
    * ```isLeapYear```
    * ```isSame```
    * ```isBefore```
    * ```isAfter```
    * ```isBetween```
* new plugin ```relativeTimePlugin```
    * ```timeAgo```
    * ```timeIn```

### Version 0.6.1

* minor bug fixes and corrections

### Version 0.6.0

* ```countdown``` re-created as a plugin to keep the core api size small.
    * instead of entering an hour, minute, second, you now add a date instead.
* ```addPlugin```
    * now accepts an array of plugins in the first parameter
* ```uptoDate.plugins``` array which holds references to any plugins installed for uptoDate
* timezone support for ```getLocalDate``` and ```getLocalTime```
* ```add```/```subtract```/```modify```
    * now you can use 'day' || 'days' etc instead of just 'days'
    * years, months and weeks can now be added or subtracted
* ```dateString``` has been renamed to ```getLocalDate```
* ```timeString``` has been renamed to ```getLocalTime```
* minor clean up of uptoDate.js, Changelog and Readme

### Version 0.5.0

* created an extendable plugin system so the core api stays small.
    * example template inside the plugin folder
* ```addPlugin```
    * plugins to add or overide uptoDate methods
* created time tracker as an optional plugin instead of removing

### Version 0.4.0

* ```add```
    * days, hours, minutes, seconds, milliseconds
* ```subtract```
    * days, hours, minutes, seconds, milliseconds
* ```modify```
    * modify allows adding, subtracting, multiplying and dividing dates and times
* depreciating tracker functions before version 1.0.0
    * should be remade as an example or plugin instead.

### Version 0.3.0

* ```pause```
    * new pause function for the tracker
* tracker now outputs time paused with overall time
    * 0 days, 2 hours, 0 minutes, 0 seconds
* ```isToday```
    * returns boolean to check if entered date is today
* ```getDayOfYear```
    * returns day since the start of the year
* ```formatDate```
    * dayOfYear added
* replaced the Array.find methods with Array.forEach for IE 11 support

### Version 0.2.0

* ```setLocale```
    * chainable
* ```formatDate```
    * new objects -
    * iso
    * utc
    * full
* ```cloneDate```
    * immutable clone

### Version 0.1.0

* local date
* local time
* day
* month
* format date
* format time
* time tracker
* countdown
* time ago
* time in


js/main.js
/**
 * Basic Example
 */

// start the utDate
const utDate = new uptoDate();
utDate.addPlugin([timeTrackerPlugin, countdownPlugin, queryPlugin, relativeTimePlugin, startOfPlugin]);

const start = utDate.start();

// init 5 minutes for timeIn
const newDateIn = new Date();
const dateIn = utDate.add('minutes', 5, newDateIn);

// get ids to populate
// title tag
let title = document.getElementsByTagName('title')[0];

// on mouse over events for title, date and time ids
// title
let titleID = document.getElementById('title');
// title mouseover/out events
titleID.addEventListener("mouseover", function(event) {
    event.target.textContent = 'const utDate = new uptoDate();';
}, false);
titleID.addEventListener("mouseout", function(event) {
    event.target.textContent = 'Up To Date';
}, false);

// date
let dateID = document.getElementById('date');
// date mouseover/out events
dateID.addEventListener("mouseover", function(event) {
    event.target.textContent = 'utDate.getLocalDate();';
}, false);
dateID.addEventListener("mouseout", function(event) {
    event.target.textContent = utDate.getLocalDate();
}, false);

// time
let timeID = document.getElementById('time');
// time mouseover/out events
timeID.addEventListener("mouseover", function(event) {
    event.target.textContent = 'utDate.getLocalTime();';
}, false);
timeID.addEventListener("mouseout", function(event) {
    event.target.textContent = utDate.getLocalTime();
}, false);

let startID = document.getElementById('start');
let agoID = document.getElementById('ago');
let inID = document.getElementById('in');
let stopID = document.getElementById('stop');
let timeTID = document.getElementById('timeOnSite');
let countdownID = document.getElementById('countdown');
let buttonID = document.getElementById('toggle');

// helper checks
let isStopped = false;
let timeinCheck = false;

// set date & time
dateID.textContent = utDate.getLocalDate();
timeID.textContent = utDate.getLocalTime();

// start time
startID.textContent = 'Start Time: ' + utDate.getLocalTime();

// update current time every second
setInterval(function(){
 
    // update title
    title.textContent = 'uptoDate ' + utDate.getLocalTime();

    if(!isStopped) {
        let update = utDate.update();
        timeTID.textContent = update.string;
    }
    // 5 minute countdown example
    const countdown = utDate.countdown(dateIn);
    countdownID.textContent = countdown.string;

    // time ago
    const timeago = utDate.timeAgo(start);
    agoID.textContent = timeago;

    // time in
    const timein = utDate.timeIn(dateIn);
    if(timein === 'Now' && !timeinCheck) {
        timeinCheck = true;
        inID.textContent = 'Now';
    }
    else if(!timeinCheck) {
        inID.textContent = timein;
    }


}, 1000);

// button click
buttonID.addEventListener('click', function(){
    if(isStopped) {
        this.textContent = 'STOP';
        utDate.start();
        startID.textContent = 'Start Time: ' + utDate.getLocalTime();
        stopID.textContent = 'End Time: ';
        isStopped = false;
    }
    else {
        this.textContent = 'START';
        isStopped = true;
        utDate.stop();
        stopID.textContent = 'End Time: ' + utDate.getLocalTime();
    }
})

plugins/queryPlugin/queryPlugin.js
/**
 * uptoDate query plugin  
 * @param {*} uptoDate - the uptoDate class
 * @param {*} options - optional plugin options
 */
'use strict'

// name of plugin
function queryPlugin(uptoDate, options) {

    /**
     * isPast     
     * Helper function to check if entered date is in the past
     * @method uptoDate.isPast
     * @type {function}
     * @param {*} date - date to check against
     * @returns boolean
     */
    uptoDate.prototype.isPast = function(date) {
        const todayDate = new Date();
        const newDate = this.cloneDate(date);

        const isPast = newDate - todayDate;

        if(isPast < 0) { return true }
        else { return false; }

    }


    /**
     * isFuture     
     * Helper function to check if entered date is in the future
     * @method uptoDate.isFuture
     * @type {function}
     * @param {*} date - date to check against
     * @returns boolean
     */
    uptoDate.prototype.isFuture = function(date) {
        const todayDate = new Date();
        const newDate = this.cloneDate(date);

        if(newDate > todayDate) { return true; }
        else { return false; }

    }


    /**
     * isYearMonthDay     
     * Helper function to check if the two dates entered are the same year, month and day.
     * Used internally for isToday, isTomorrow and isYesterday
     * @method uptoDate.isFuture
     * @type {function}
     * @param {*} date - date to check against
     * @returns object with year, month and day function which return a boolean
     */
    uptoDate.prototype.isYearMonthDay = function(newDate, todayDate) {
        return {
            year: function() {
                if(newDate.getFullYear() === todayDate.getFullYear()) { return true; }
                else { return false; }
            },
            month: function() {
                if(newDate.getMonth() === todayDate.getMonth()) { return true; }
                else { return false; }
            },
            day: function() {
                if(newDate.getDate() === todayDate.getDate()) { return true; }
                else { return false; }
            }
        }
    }

    
    /**
     * isToday     
     * Helper function to check if entered date is today
     * @method uptoDate.isToday
     * @type {function}
     * @param {*} date - date to check against
     * @returns boolean
     */
    uptoDate.prototype.isToday = function(date) {
        const todayDate = new Date();
        const newDate = this.cloneDate(date);

        const isToday = this.isYearMonthDay(newDate, todayDate);

        if(isToday.year() && isToday.month() && isToday.day()) { return true; }
        else { return false; }
    }


    /**
     * isTomorrow     
     * Helper function to check if entered date is tomorrow
     * @method uptoDate.isTomorrow
     * @type {function}
     * @param {*} date - date to check against
     * @returns boolean
     */
    uptoDate.prototype.isTomorrow = function(date) {
        const tomorrow = this.add('day', 1);
        const newDate = this.cloneDate(date);

        const isTomorrow = this.isYearMonthDay(newDate, tomorrow);
    
        //console.log(newd.day());
        if(isTomorrow.year() && isTomorrow.month() && isTomorrow.day()) { return true; }
        else { return false; }
    }


    /**
     * isYesterday     
     * Helper function to check if entered date was yesterday
     * @method uptoDate.isYesterday
     * @type {function}
     * @param {*} date - date to check against
     * @returns boolean
     */
    uptoDate.prototype.isYesterday = function(date) {
        const yesterday = this.subtract('day', 1);
        const newDate = this.cloneDate(date);

        const isYesterday = this.isYearMonthDay(newDate, yesterday)
        if(isYesterday.year() && isYesterday.month() && isYesterday.day()) { return true; }
        else { return false; }
    }


    /**
     * isWeekend      
     * Helper function to check date entered is a weekend  
     * Warning: This checks to see if the date is a Saturday or Sunday.
     * Some countries only have one day weekends or start the weekend on a Friday.
     * @method uptoDate.isWeekend
     * @type {function}
     * @param {*} date - date to check against
     * @returns boolean
     */
    uptoDate.prototype.isWeekend = function(date) {
        const newDate = this.cloneDate(date);
        const day = newDate.getDay();

        // check for saturday or sunday
        // 0 = sunday
        if(day === 6 || day === 0) {
            return true;
        }
        else { return false; }

    }


    /**
     * isLeapYear      
     * Helper function to check if the year from the entered date is a leap year
     * @method uptoDate.isLeapYear
     * @type {function}
     * @param {*} date - date to check against
     * @returns boolean
     */
    uptoDate.prototype.isLeapYear = function(date) {
        const newDate = this.cloneDate(date);
        const dateYear = newDate.getFullYear();

        // check if leap year depending on if its the 29th of feb
        if(new Date(dateYear, 1, 29).getMonth() === 1) {
            return true;
        }
        else { return false; }

    }


    /**
     * isSame      
     * Helper function to check if the first date is the same as the second date
     * @method uptoDate.isSame
     * @type {function}
     * @param {*} first - date to check if before second date
     * @param {*} second - date to check against
     * @returns boolean
     */
    uptoDate.prototype.isSame = function(first, second) {

        if(first === second) { return true; }
        else { return false; }
    }


    /**
     * isBefore      
     * Helper function to check if the first date is before the second date
     * @method uptoDate.isBefore
     * @type {function}
     * @param {*} first - date to check if before second date
     * @param {*} second - date to check against
     * @returns boolean
     */
    uptoDate.prototype.isBefore = function(first, second) {

        if(first < second) { return true; }
        else { return false; }
    }


    /**
     * isAfter       
     * Helper function to check if the first date is after the second date
     * @method uptoDate.isAfter
     * @type {function}
     * @param {*} first - date to check if after second date
     * @param {*} second - date to check against
     * @returns boolean
     */
    uptoDate.prototype.isAfter = function(first, second) {

        if(first > second) { return true; }
        else { return false; }
    }


    /**
     * isBetween       
     * Helper function to check if the date between two dates
     * @method uptoDate.isBetween
     * @type {function}
     * @param {*} date - date to check against
     * @param {*} dateBetweenMin - min date to check against
     * @param {*} dateBetweenMax - max date to check against
     * @returns boolean
     */
    uptoDate.prototype.isBetween = function(date, dateBetweenMin, dateBetweenMax) {

        if(date > dateBetweenMin && date < dateBetweenMax) { return true; }
        else { return false; }
    }

}

plugins/startOfPlugin/startOfPlugin.js
/**
 * uptoDate start of plugin
 * set the date to the start of the day, week, month, year
 * @param {*} uptoDate - the uptoDate class
 * @param {*} options - optional plugin options
 */
'use strict'

// name of plugin
function startOfPlugin(uptoDate, options) {

    /**
     * startOf  
     * sets the date to the start of x
     * @method uptoDate.startOfDay
     * @type {function}
     * @param {string} string - string to enter. 'day', 'week', 'month' or 'year'
     * @param {*} [date='new Date()'] - optional input of date.
     * @returns returns date with entered start of
     */
    uptoDate.prototype.startOf = function(string, date) {
        if(date === undefined || date === null) { date = new Date(); }

        const newDate = this.cloneDate(date);

        if(string === 'day') { return this.startOfDay(newDate) }
        else if(string === 'week') { return this.startOfWeek(newDate) }
        else if(string === 'month') { return this.startOfMonth(newDate) }
        else if(string === 'year') { return this.startOfYear(newDate) }
    }
    

    /**
     * startOfDay  
     * sets the date entered or today as midnight 
     * @method uptoDate.startOfDay
     * @type {function}
     * @param {*} [date='new Date()'] - optional input of date.
     * @returns returns date entered or today at midnight (start of day)
     */
    uptoDate.prototype.startOfDay = function(date) {
        if(date === undefined || date === null) { date = new Date(); }

        const newDate = this.cloneDate(date);
        newDate.setHours(0, 0, 0, 0);

        return newDate;

    }


    /**
     * startOfWeek   
     * sets the date entered to the start of the week 
     * @method uptoDate.startOfWeek
     * @type {function}
     * @param {*} [date='new Date()'] - optional input of date.
     * @param {number} [weekStart=1] - optional day to start the week, default is Monday
     * @returns returns date entered start of week
     */
    uptoDate.prototype.startOfWeek = function(date, weekStart) {
        if(date === undefined || date === null) { date = new Date(); }
        if(weekStart === undefined || weekStart === null) { weekStart = 1; }

        const newDate = this.cloneDate(date);
        const day = newDate.getDay();

        const startOfWeek = newDate.setDate(newDate.getDate() - day + weekStart);
        
        return this.startOfDay(new Date(startOfWeek));

    }


    /**
     * startOfMonth  
     * sets the date entered to the start of the month 
     * @method uptoDate.startOfMonth
     * @type {function}
     * @param {*} [date='new Date()'] - optional input of date.
     * @returns returns date entered start of month
     */
    uptoDate.prototype.startOfMonth = function(date) {
        if(date === undefined || date === null) { date = new Date(); }

        const newDate = this.cloneDate(date);
        const currentDayInMonth = newDate.getDate();

        const diff = currentDayInMonth - 1;
        const startOfMonth = newDate.setDate(currentDayInMonth - diff);
        
        return this.startOfDay(new Date(startOfMonth));

    }


    /**
     * startOfYear 
     * sets the date entered to the start of the year 
     * @method uptoDate.startOfYear
     * @type {function}
     * @param {*} [date='new Date()'] - optional input of date.
     * @returns returns date entered start of year
     */
    uptoDate.prototype.startOfYear = function(date) {
        if(date === undefined || date === null) { date = new Date(); }

        const newDate = this.cloneDate(date);
        // set date to January
        const setMonth = newDate.setMonth(0);

        // set the date to the first day in the month
        const startOfMonth = this.startOfMonth(newDate);

        return this.startOfDay(new Date(startOfMonth));

    }
}

plugins/countdownPlugin/countdownPlugin.js
/**
 * uptoDate countdown plugin
 * Helper method when creating a countdown.
 * @param {*} uptoDate - the uptoDate class
 * @param {*} options - optional plugin options
 */
'use strict'

function countdownPlugin(uptoDate, options) {

    // check to see if the uptoDate countdown is finished
    uptoDate.countdownFinished = false;

    /**
     * countdown    
     * Simple countdown between two dates
     * @method uptoDate.countdown
     * @type {function}
     * @param {*} date - date to countdown to
     * @param {*} [fromDate=Date()] - optional date to countdown from
     */
    uptoDate.prototype.countdown = function(date, fromDate) {

        if(fromDate === undefined || fromDate === null) { fromDate = new Date(); }

        const start = this.cloneDate(fromDate)
        // get new date and set time from paramters
        const countdownDate = this.cloneDate(date);

        // get the difference between the start of the countdown and the end time
        const td = this.getTimeDifference(start, countdownDate);

        // countdown finished
        if(td.days === 0 && td.hours === 0 && td.minutes === 0 && td.seconds === 0) {
            this.countdownFinished = true;
        }

        if(!this.countdownFinished) {
            const countdownObj = {
                time: td,
                string: td.days + ' Days ' + td.hours + ' Hours ' + td.minutes + ' Minutes ' + td.seconds + ' Seconds '
            }
            return countdownObj;
        }
        else {
            const countdownFinishedObj = {
                string: 'Countdown Finished!',
            }
            
            return countdownFinishedObj;
        }
        
    }
}

js/uptoDate.js
/**
 * @author Conor Irwin <https://github.com/RetroVX> 
 * @license {@link http://opensource.org/licenses/MIT|MIT License}
 * uptoDate
 * A tiny time and date helper 'library' 
 * @version 0.1.0
 */
function uptoDate() {

    this.dateOptions = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' };
    this.today = new Date();

    // get time started and time stopped
    this.timeStarted;
    this.timeCurrent;

    // check to see if the uptoDate countdown is finished
    this.countdownFinished = false;

}


/**
 * dateString  
 * @method uptoDate.dateString
 * @type {function}
 * @param {*} [date='new Date()'] - optional input of date.
 * @param {string} [locale='default'] - optional input of locale. Eg: 'en-US'
 * @returns return the local date string for date entered
 */
uptoDate.prototype.dateString = function(date, locale) {
    if(date === undefined || date === null) { date = new Date(); };
    if(locale === undefined || locale === null) { locale = 'default' };

    return date.toLocaleDateString(locale, this.dateOptions); 
}


/**
 * timeString  
 * @method uptoDate.timeString
 * @type {function}
 * @param {*} [date='new Date()'] - optional input of date.
 * @param {string} [locale='default'] - optional input of locale. Eg: 'en-US'
 * @returns return the local time string from date entered 
 */
uptoDate.prototype.timeString = function(date, locale) {
    if(date === undefined || date === null) { date = new Date(); };
    if(locale === undefined || locale === null) { locale = 'default' };


    return date.toLocaleTimeString(locale); 
}


/**
 * formatDay  
 * @method uptoDate.formatDay
 * @type {function}
 * @param {*} [date='new Date()'] - optional input of date.
 * @param {boolean} [short=false] - optional output of short day (3 letters, Sunday -> Sun)
 * @returns return day eg Monday
 */
uptoDate.prototype.formatDay = function(date, short) {
    if(date === undefined || date === null) { date = new Date(); };
    if(short === undefined || short === null) { short = false };

    const day = date.getDay();
    const dayArray = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];

    const findDay = dayArray.find(function(d, index) {
        if(index === day) { return d };
    })

    if(short) { return findDay.slice(0, 3); }
    else { return findDay; }
}


/**
 * formatMonth 
 * @method uptoDate.formatMonth
 * @type {function}
 * @param {*} [date='new Date()'] - optional input of date.
 * @param {boolean} [short=false] - optional output of short month (3 letters, January -> Jan)
 * @returns return month eg June
 */
uptoDate.prototype.formatMonth = function(date, short) {
    if(date === undefined || date === null) { date = new Date(); };
    if(short === undefined || short === null) { short = false };

    const month = date.getMonth();
    const monthArray = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

    const findMonth = monthArray.find(function(m, index) {
        if(index === month) { return m };
    });

    if(short) { return findMonth.slice(0, 3); }
    else { return findMonth; }
}


/**
 * formatDate  
 * @method uptoDate.formatDate
 * @type {function}
 * @param {*} [date='new Date()'] - optional input of date.
 * @returns return object with day, month, full year and short year
 */
uptoDate.prototype.formatDate = function(date) {
    if(date === undefined || date === null) { date = new Date(); };

    // convert full year to short year
    // 2019 -> 19
    const convertFullYear = date.getFullYear().toString();
    const shortYear = convertFullYear.slice(2, 4);

    const dateObj = {
        day: date.getDate(),
        fullDay: this.formatDay(date),
        shortDay: this.formatDay(date, true),
        month: date.getMonth(),
        shortMonth: this.formatMonth(date, true),
        fullMonth: this.formatMonth(date),
        fullYear: date.getFullYear(),
        shortYear: shortYear
    }

    return dateObj;
}


/**
 * formatTime  
 * @method uptoDate.formatTime
 * @type {function}
 * @param {*} [date='new Date()'] - optional input of date.
 * @returns return object with hour, minute, second and milliseconds of date
 */
uptoDate.prototype.formatTime = function(date) {
    if(date === undefined || date === null) { date = new Date(); };

    const timeObj = {
        hours: date.getHours(),
        minutes: date.getMinutes(),
        seconds: date.getSeconds(),
        ms: date.getMilliseconds()
    }

    return timeObj;
}


/**
 * start  
 * Get the time now, if started task
 * @method uptoDate.start
 * @type {function}
 * @returns {number} returns Date().getTime();
 */
uptoDate.prototype.start = function() {

    // update timestarted with current time
    this.timeStarted = new Date().getTime();

    return this.timeStarted;
    
}


/**
 * update  
 * Get time updated so we can get difference between start and current
 * @method uptoDate.update
 * @type {function}
 * @returns {object} - returns the updated time and date object
 */
uptoDate.prototype.update = function() {

    // get date
    const date = new Date();

    // update time stopped with current time
    this.timeCurrent = date.getTime();

    // get the time difference between start and stop
    const difference = this.getTimeDifference();

    const timeObj = {
        date: date,
        time: difference,
        string: difference.days + ' Days ' + difference.hours + ' Hours ' + difference.minutes + ' Minutes ' + difference.seconds + ' Seconds '
    }
    
    return timeObj;

}


/**
 * stop  
 * run update and return the date and time. 
 * @method uptoDate.stop
 * @type {function} 
 * @returns {object} - returns the stopped time and date object
 */
uptoDate.prototype.stop = function() {
    const stopObj = this.update();
    this.timeStarted = null;
    this.timeCurrent = null;

    return stopObj;
}


/**
 * getTimeDifference  
 * get the time difference between start and stop times
 * @method uptoDate.getTimeDifference
 * @type {function}
 * @param {*} start - start time from date.getTime()
 * @param {*} stop - stop time from date.getTime()
 */
uptoDate.prototype.getTimeDifference = function(start, stop) {

    if(start === undefined || start === null) { start = this.timeStarted };
    if(stop === undefined || stop === null) { stop = this.timeCurrent };

    // get total seconds between the times
    let delta = Math.abs(start - stop) / 1000;

    // calculate (and subtract) whole days
    let days = Math.floor(delta / 86400);
    delta -= days * 86400;

    // calculate (and subtract) whole hours
    let hours = Math.floor(delta / 3600) % 24;
    delta -= hours * 3600;

    // calculate (and subtract) whole minutes
    let minutes = Math.floor(delta / 60) % 60;
    delta -= minutes * 60;

    // what's left is seconds
    let seconds = delta % 60; 

    const timeObj = {
        days: days,
        hours: hours,
        minutes: minutes,
        seconds: Math.floor(seconds)
    }

    return timeObj;
}


/**
 * countdown  
 * Simple countdown for hours/minutes/seconds
 * @method uptoDate.countdown
 * @type {function}
 * @param {number} hours - the hour of day you want the countdown to end
 * @param {number} minutes - the minute you want the countdown to end
 * @param {number} seconds - the second you want the countdown to end
 * @example
 * // countdown to 8pm
 * .countdown(20, 0, 0);
 */

uptoDate.prototype.countdown = function(hours, minutes, seconds) {
    // get time now
    const start = new Date().getTime();
    // get new date and set time from paramters
    const countdownDate = new Date();
    const countdown = countdownDate.setHours(hours, minutes, seconds);

    // get the difference between the start of the countdown and the end time
    const td = this.getTimeDifference(start, countdown);

    // countdown finished
    if(td.days === 0 && td.hours === 0 && td.minutes === 0 && td.seconds === 0) {
        this.countdownFinished = true;
    }

    if(!this.countdownFinished) {
        const countdownObj = {
            time: td,
            string: td.days + ' Days ' + td.hours + ' Hours ' + td.minutes + ' Minutes ' + td.seconds + ' Seconds '
        }
        return countdownObj;
    }
    else {
        const countdownFinishedObj = {
            string: 'Countdown Finished!',
        }
        
        return countdownFinishedObj;
    }
    
}


/**
 * timeAgo    
 * get a string with 'x time ago' eg: 5 minutes ago
 * @method uptoDate.timeAgo
 * @type {function}
 * @param {*} time - time to check against (requires Date().getTime() format)
 */
uptoDate.prototype.timeAgo = function(time) {
    // get time now
    const date = new Date().getTime();

    // get the difference between now and time entered
    const td = this.getTimeDifference(time, date);

    let timeago;

    // update depending on time passed
    // checks to see if the time is at 1 so it does not display 1 seconds ago etc
    if(td.days === 1) { timeago = td.days + ' day ago'; }
    else if(td.days > 1) { timeago = td.days + ' days ago'; }
    else if(td.hours === 1) { timeago = td.hours + ' hour ago'; }
    else if(td.hours > 1) { timeago = td.hours + ' hours ago'; }
    else if(td.minutes === 1) { timeago = td.minutes + ' minute ago'; }
    else if(td.minutes > 1) { timeago = td.minutes + ' minutes ago'; }
    else if(td.seconds === 1) { timeago = td.seconds + ' second ago'; }
    else if(td.seconds > 1) { timeago = td.seconds + ' seconds ago'; }
    else { timeago = 'Just now'}

    return timeago;
}


/**
 * timeIn     
 * get a string with 'in x time' eg: in 5 minutes
 * @method uptoDate.timeIn
 * @type {function}
 * @param {*} time - time to check against (requires Date().getTime() format)
 */
uptoDate.prototype.timeIn = function(time) {
    // get time now
    const date = new Date().getTime();

    // get the difference between now and time entered
    const td = this.getTimeDifference(date, time);

    let timein;

    // update depending on time passed
    // checks to see if the time is at 1 so it does not display in 1 seconds etc
    if(td.days === 1) { timein = 'in ' + td.days + ' day'; }
    else if(td.days > 1) { timein = 'in ' + td.days + ' days'; }
    else if(td.hours === 1) { timein = 'in ' +  td.hours + ' hour'; }
    else if(td.hours > 1) { timein = 'in ' + td.hours + ' hours'; }
    else if(td.minutes === 1) { timein = 'in ' + td.minutes + ' minute'; }
    else if(td.minutes > 1) { timein = 'in ' + td.minutes + ' minutes'; }
    else if(td.seconds === 1) { timein = 'in ' + td.seconds + ' second'; }
    else if(td.seconds > 1) { timein = 'in ' + td.seconds + ' seconds'; }
    else { timein = 'Now'}

    return timein;
}